### PR TITLE
feat: implement custom titlebar matching editor style

### PR DESCRIPTION
**Pull request by panekj**

### CODE REVIEW

**Comment by dzhou121** [comment]
If possible, maybe we should draw the "minimize", "maximise", and "close" icons like a normal windows APP with title.

**Comment by dzhou121** [comment]
image

something like this

**Comment by panekj** [comment]
Normal window:
image
Maximised window:
image
Full titlebar:
image


@dzhou121 few things:

- [ ] do we want lapce logo in titlebar? current position is ok? same for colour? which platforms will have logo? for example VSCode has logo on Windows, but not on macOS due to window controls on left side

- [ ] do we want title? what should title have? should title be same as "os" title (currently it's workspace/folder name)

- [ ] do we want window controls on windows to match perfectly? standard Windows window controls have big padding between each other. they also look a bit blurred(?)
image


Example of draggable lapce:

https://user-images.githubusercontent.com/50457605/176743017-73656470-f7fa-4fda-9246-50dc7e57a162.mp4

**Comment by dzhou121** [comment]
>do we want lapce logo in titlebar? current position is ok? same for colour? which platforms will have logo? for example VSCode has logo on Windows, but not on macOS due to window controls on left side

Lapce logo only for Windows.

>do we want title? what should title have? should title be same as "os" title (currently it's workspace/folder name)

Probably yes for Windows only. 

>do we want window controls on windows to match perfectly? standard Windows window controls have big padding between each other. they also look a bit blurred(?)

Yes please if that's not too much of trouble.

**Comment by panekj** [comment]
image

**Comment by panekj** [comment]
The only thing left is to add dynamic title to show workspace name but I haven't figured that out yet. PR can be merged and I'll push fix for title later.
Also, restoring window currently doesn't work from window control, it works on double click on titlebar though.

**Comment by dzhou121** [comment]
Thanks!

I'm perfectly happy with it for not showing project name, because it's already at the very left. 

The restore thing might be a Druid thing. I can look into that.

File: lapce-ui/src/app.rs [hunk 1]
```diff
@@ -83,6 +83,15 @@ where
         .title(LocalizedString::new("Lapce").with_placeholder("Lapce"))
         .window_size(size)
         .set_position(pos);
+
+    #[cfg(target_os = "windows")]
+    if Config::load(&lapce_data::data::LapceWorkspace::default())
```
Comment (dzhou121) [inline lapce-ui/src/app.rs:88]: You can pass the Config in from the parameter.

File: lapce-ui/src/title.rs [hunk 2]
```diff
@@ -97,7 +105,19 @@ impl Widget<LapceWindowData> for Title {
                     )
                 }
             }
-            _ => {}
+            _ => {
+                #[cfg(target_os = "windows")]
+                {
+                    ctx.submit_command(
```
Comment (dzhou121) [inline lapce-ui/src/title.rs:111]: Is this for when the ```custom_titlebar``` is toggled in settings? It's better to put in Update and run this when the settings is actually changed.

Comment (panekj) [reply lapce-ui/src/title.rs]: moved to `update`

File: lapce-ui/src/app.rs (hunk 1 above)
Comment (panekj) [reply lapce-ui/src/app.rs]: done

File: lapce-ui/src/title.rs [hunk 3]
```diff
@@ -126,7 +148,15 @@ impl Widget<LapceWindowData> for Title {
...
-        Size::new(bc.max().width, 28.0)
+        #[cfg(not(target_os = "windows"))]
+        {
+            Size::new(bc.max().width, 28.0)
+        }
+
+        #[cfg(target_os = "windows")]
+        {
+            Size::new(bc.max().width, 32.0)
+        }
```
Comment (panekj) [inline lapce-ui/src/title.rs:159]: Changed titlebar height on Windows to match most other apps and improve readability slightly

**Pull request merged**